posthog/models/product_intent/product_intent.py

```diff
@@ -9,9 +40,75 @@ class ProductIntent(UUIDModel):
     updated_at = models.DateTimeField(auto_now=True)
     product_type = models.CharField(max_length=255)
     onboarding_completed_at = models.DateTimeField(null=True, blank=True)
+    activated_at = models.DateTimeField(
+        null=True,
+        blank=True,
+        help_text="The date the org completed activation for the product. Generally only used to know if we should continue updating the product_intent row.",
+    )
+    activation_last_checked_at = models.DateTimeField(
+        null=True,
+        blank=True,
+        help_text="The date we last checked if the org had completed activation for the product.",
+    )
 
     class Meta:
         unique_together = ["team", "product_type"]
 
     def __str__(self):
         return f"{self.team.name} - {self.product_type}"
+
+    def has_activated_data_warehouse(self) -> bool:
+        insights = Insight.objects.filter(
+            team=self.team,
+            created_at__gte=datetime(2024, 6, 1, tzinfo=UTC),
+            query__kind="DataVisualizationNode",
+        )
+
+        excluded_tables = ["events", "persons", "sessions", "person_distinct_ids"]
+        for insight in insights:
+            if insight.query and insight.query.get("source", {}).get("query"):
+                query_text = insight.query["source"]["query"].lower()
+                # Check if query doesn't contain any of the excluded tables after 'from'
+                has_excluded_table = any(f"from {table}" in query_text.replace("\\", "") for table in excluded_tables)
```

Comment:
I don't understand this part. Why are these tables excluded? Are they just the stock ones provided by PostHog? So if not excluded, then it's assumed to be one of their tables?

This could be optimized so it's all done in the query:
```
    excluded_tables = ["events", "persons", "sessions", "person_distinct_ids"]
    
    # Create a list of Q objects for each excluded table
    excluded_patterns = [
        Q(query__source__query__icontains=f"from {table}") 
        for table in excluded_tables
    ]
    
    # Combine all exclusion patterns with OR operator
    exclusion_filter = reduce(operator.or_, excluded_patterns)
    
    # Single optimized query that:
    # 1. Filters by team and date
    # 2. Filters by query kind
    # 3. Excludes queries containing any excluded table
    # 4. Checks if any matching record exists
    return Insight.objects.filter(
        team=self.team,
        created_at__gte=datetime(2024, 6, 1, tzinfo=UTC),
        query__kind="DataVisualizationNode",
    ).exclude(
        exclusion_filter
    ).filter(
        query__source__query__isnull=False
    ).exists()
    ```

Reply:
This query is what Eric gave me for data warehouse activation. I'll look at your optimization! I haven't tested that this works at all yet haha.